invenio_oauthclient/contrib/globus.py

```diff
@@ -0,0 +1,226 @@
+   full documentation on all app fields, see:
+   `https://docs.globus.org/api/auth/developer-guide/`
+
+4. Grab the *Client ID* and *Client Secret* after registering the application
+   and add them to your instance configuration (``invenio.cfg``):
+
+   .. code-block:: python
+
+        GLOBUS_APP_CREDENTIALS = dict(
+            consumer_key='<CLIENT ID>',
+            consumer_secret='<CLIENT SECRET>',
+        )
+
+5. Now go to your site: http://localhost:5000/oauth/authorized/globus/
+
+6. You should see Globus listed under Linked accounts:
+   http://localhost:5000/account/settings/linkedaccounts/
+
+"""
+
+from flask import current_app, redirect, url_for
+from flask_login import current_user
+from invenio_db import db
+
+from invenio_oauthclient.errors import OAuthResponseError
+from invenio_oauthclient.models import RemoteAccount
+from invenio_oauthclient.utils import oauth_link_external_id, \
+    oauth_unlink_external_id
+
+
+REMOTE_APP = dict(
+    title='Globus',
+    description='Research data management simplified.',
+    icon='',
+    authorized_handler='invenio_oauthclient.handlers'
+                       ':authorized_signup_handler',
+    disconnect_handler='invenio_oauthclient.contrib.globus'
+                       ':disconnect_handler',
+    signup_handler=dict(
+        info='invenio_oauthclient.contrib.globus:account_info',
+        setup='invenio_oauthclient.contrib.globus:account_setup',
+        view='invenio_oauthclient.handlers:signup_handler',
+    ),
+    params=dict(
+        request_token_params={'scope': 'openid email profile'},
+        base_url='https://auth.globus.org/',
+        request_token_url=None,
+        access_token_url='https://auth.globus.org/v2/oauth2/token',
+        access_token_method='POST',
+        authorize_url='https://auth.globus.org/v2/oauth2/authorize',
+        app_key='GLOBUS_APP_CREDENTIALS',
+    )
+)
+"""Globus remote application configuration."""
+
+GLOBUS_USER_INFO_URL = 'https://auth.globus.org/v2/oauth2/userinfo'
+GLOBUS_USER_ID_URL = 'https://auth.globus.org/v2/api/identities'
+GLOBUS_EXTERNAL_METHOD = 'globus'
+
+
+def get_dict_from_response(response):
+    """Check for errors in the response and return the resulting JSON"""
+    if getattr(response, '_resp') and response._resp.code > 400:
+        raise OAuthResponseError(
+                'Application mis-configuration in Globus', None, response
+            )
+
+    return response.data
+
+
+def get_user_info(remote):
+    """Get user information from Globus.
+
+    See the docs here for v2/oauth/userinfo:
+    https://docs.globus.org/api/auth/reference/
+    """
+    response = remote.get(GLOBUS_USER_INFO_URL)
+    user_info = get_dict_from_response(response)
+    username = response.data['preferred_username']
+    if '@' in username:
+        user_info['preferred_username'], _ = username.split('@')
+    return user_info
+
+
+def get_user_id(remote, email):
```

Comment:
Invenio binds user ids to usernames, so the Globus ID is also needed. This results in this second required call. The information _is_ stored within the id token, but one of the advantages to hitting the REST APIs directly in this case is we don't require any other dependencies to look inside the OIDC token or the Globus API.

Reply:
Is it potentially problematic that Globus Auth explicitly states that the mapping between identity username and ID is not fixed over time?
https://docs.globus.org/api/auth/reference/#get_v2_api_identities_usernames_lt_list_of_identity_names_gt

Reply:
I didn't think it was possible for one identity username to map to multiple ids. The example shows calling in with multiple identity usernames and returning a bunch of corresponding ids to those usernames. 

For mapping identity usernames to IDs? That's a good question. If the identity username were to change hands, would the ID also change to something else thereby invalidating the user account on Invenio? Something I'm trying to confirm is that users in Invenio are tracked by external provider id rather than by username.